### PR TITLE
HPCC-27581 Remove ThreadDestroyListLock

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
Pushed to testing and for discussion.

**Comment by richardkchapman** [comment]
The core needs investigating. And I'm not sure this should target a point release

**Comment by ghalliday** [comment]
I agree with both comments.  I was lazy about rebasing before pushing.

**Comment by ghalliday** [comment]
I understand the problem - and possibly why the code was there in the first place.  If nothing else I can now improve the comment. to make it clear.

**Comment by ghalliday** [comment]
Actually, I think it is revealing a potential bug....

File: system/jlib/jthread.cpp [hunk 1]
```diff
@@ -119,10 +119,40 @@ void disableThreadSEH() { SEHHandling=false; } // only prevents new threads from
...
+  The joining thread checks it can get that critical section before continuing.  The joining thread may wake up, try
+  and get the critical section, and then block because the other thread hasn't released it yet, but this is better than
+  the previous situation which would spin-loop.  The join will be blocked by all terminating threads - in a signal,
+  rather than purely user code.
+  [This is the new implementation]
 
+* Terminating thread conditionally signals.
+  The terminating thread could call Release(), and only signal() the semaphore if the object was not destroyed.  This would cope
+  with the common joining pattern, but would fail if another object held onto a reference to the thread but didn't call join().
+  Potentially more efficient, but more scope for accidental bugs.
```
Comment (jakesmith) [inline system/jlib/jthread.cpp:153]: Existing code was very confusing (double Release) etc, and wasteful list, looks a lot better.

I've eventually convinced myself of the need for the CS to be global, but it's a bit of a pity that still a lock involved in thread handling as they finished up/joined.

If startRelease's usages could be removed, and join() mandated, then we could avoid _threadmain ever having the possibility of destroying the thread (get rid of start linking and _threadmain releasing), and then we wouldn't need the CS.

**Review by jakesmith — APPROVED** [review]
@ghalliday - looks good, just some musings re. whether we can get rid of global CS, startRelease and mandate join(), but separate JIRA if so.

File: system/jlib/jthread.cpp [hunk 2]
```diff
@@ -475,22 +492,10 @@ bool Thread::join(unsigned timeout)
...
-        if (st==10)     
-            PROGLOG("Thread::join race");   
-#endif
-        Sleep(st);      // switch back to exiting thread (not very elegant!)
-        st++;
-        if (st>10)
-            st = 10; // note must be non-zero for high priority threads
+
+    {
+        CriticalBlock block(ThreadDestroyLock);
```
Comment (mckellyln) [inline system/jlib/jthread.cpp:497]: Do we need this CB here ?

**Review by mckellyln — COMMENTED** [review]
Looks better, more direct and simple.
Just have one question.

File: system/jlib/jthread.cpp (hunk 1 above)
Comment (richardkchapman) [reply system/jlib/jthread.cpp:153]: Perhaps normal threads should mandate join() before the are destroyed (and can check for same and report if missed), and a different class (that does not have a join method) provided for "fire and forget" threads. Though I'm not really convinced there is really any such thing - at the very least they need to be "joined" at process termination.

File: system/jlib/jthread.cpp (hunk 2 above)
Comment (ghalliday) [reply system/jlib/jthread.cpp:497]: yes - the comment should probably precede the critical block.  It is preventing the other thread stalling between the signal and the release, and this thread continuing, and disposing of the object that contains the threading object.  (Which does happen...)

File: system/jlib/jthread.cpp (hunk 1 above)
Comment (ghalliday) [reply system/jlib/jthread.cpp:153]: I think threads should always have to be joined, but implement that as a future separate change.

**Comment by richardkchapman** [comment]
Can you rebase and retarget master please

File: system/jlib/jthread.cpp (hunk 2 above)
Comment (ghalliday) [reply system/jlib/jthread.cpp:501]: moved the comment, and clarified it.

**Review by mckellyln — APPROVED** [review]

**Comment by ghalliday** [comment]
@richardkchapman squashed and rebased/target 8.8.x.  Careful review recommended.

**Pull request merged**